### PR TITLE
Fix test_shader_lang (for 3.x)

### LINKED ISSUE #50975: 3.x `shaderlang` test fails
### Godot version

3.x (8f891535d04fd2efd50459f9dbd5f3f3058585ba), 3.3

### System information

Linux, Mageia 9 x86_64

### Issue description

Running `godot --test shaderlang` gives the following error in the `3.x` branch (and `3.3`):

```
ERROR: Method failed. Returning: nullptr
   at: test (main/tests/test_shader_lang.cpp:322)
```

This is one of the old-style tests that we didn't properly maintain before 4.0, so it's low priority to fix, but might still be worth doing, also to have an actual test case to check potential regressions in shader lang changes.

For `master`, this test hasn't been ported yet to the new doctest framework (the code is still there but not used).

### Steps to reproduce

`godot --test shaderlang`

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by Chaosus**
Fix #50975. Not quite sure that it's the correct fix but it fixes the error for sure

### CODE REVIEW

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Oops, used wrong button (wanted to rebase, not do a remote merge).

**Comment by akien-mga** [comment]
I tested this rebased on current `3.x` and it seems to still fail:
```
$ godot-3.x --test shaderlang
Godot Engine v3.5.beta.custom_build.47334b721 - https://godotengine.org
Found discrete GPU, setting DRI_PRIME=1 to use it.
Note: Set DRI_PRIME=0 in the environment to disable Godot from using the discrete GPU.
OpenGL ES 3.0 Renderer: AMD VEGAM (DRM 3.44.0, 5.16.10-desktop-3.mga9, LLVM 13.0.1)
Async. shader compilation: OFF
 
ERROR: Method failed. Returning: nullptr
   at: test (main/tests/test_shader_lang.cpp:347)
```

**Comment by Chaosus** [comment]
@akien-mga Ok, I've fixed it by adding a check for empty arg. Such things are better to be handled by @Xrayez - I'm not a specialist in tests..

**Comment by akien-mga** [comment]
Ah right it needs a shader of course, my bad. I'll test again both with and without arg.

**Comment by akien-mga** [comment]
Seems like it supports only spatial shaders? That's testing on a https://github.com/godotengine/godot-demo-projects 2D shader:
```
$ godot-3.x --test shaderlang 2d/screen_space_shaders/shaders/pixelize.shader 
Godot Engine v3.5.beta.custom_build.1770207bb - https://godotengine.org
Found discrete GPU, setting DRI_PRIME=1 to use it.
Note: Set DRI_PRIME=0 in the environment to disable Godot from using the discrete GPU.
OpenGL ES 3.0 Renderer: AMD VEGAM (DRM 3.44.0, 5.16.10-desktop-3.mga9, LLVM 13.0.1)
Async. shader compilation: OFF
 
tokens:

1: EOF
1: IDENTIFIER(canvas_item)
1: VARYING
3: OUT
3: TYPE_VEC2
3: IDENTIFIER(size_x)
3: OP_ASSIGN_SUB
3: REAL_CONSTANT(0.008)
3: VARYING
4: OUT
4: TYPE_VEC2
4: IDENTIFIER(size_y)
4: OP_ASSIGN_SUB
4: REAL_CONSTANT(0.008)
4: VARYING
6: TYPE_BOOL
6: IDENTIFIER(fragment)
6: COMMA
6: COLON
6: PARENTHESIS_CLOSE
7: TYPE_VEC3
7: IDENTIFIER(uv)
7: OP_ASSIGN_SUB
7: IDENTIFIER(SCREEN_UV)
7: VARYING
8: IDENTIFIER(uv)
8: OP_ASSIGN_DIV
8: IDENTIFIER(mod)
8: COMMA
8: IDENTIFIER(uv)
8: PERIOD
8: TYPE_VEC3
8: COMMA
8: IDENTIFIER(size_x)
8: PERIOD
8: IDENTIFIER(size_y)
8: COLON
8: COLON
8: VARYING
10: IDENTIFIER(COLOR)
10: IN
10: IDENTIFIER(rgb)
10: OP_ASSIGN_SUB
10: IDENTIFIER(textureLod)
10: COMMA
10: IDENTIFIER(SCREEN_TEXTURE)
10: PERIOD
10: IDENTIFIER(uv)
10: PERIOD
10: REAL_CONSTANT(0)
10: COLON
10: IN
10: IDENTIFIER(rgb)
10: VARYING
11: QUESTION

Error at line: 1: Invalid shader type. Valid types are: 'spatial'
```

Testing on a 3D shader in the TPS demo:
```
$ godot-3.x --test shaderlang enemies/red_robot/laser/LaserShader.shader 
Godot Engine v3.5.beta.custom_build.1770207bb - https://godotengine.org
Found discrete GPU, setting DRI_PRIME=1 to use it.
Note: Set DRI_PRIME=0 in the environment to disable Godot from using the discrete GPU.
OpenGL ES 3.0 Renderer: AMD VEGAM (DRM 3.44.0, 5.16.10-desktop-3.mga9, LLVM 13.0.1)
Async. shader compilation: OFF
 
tokens:

1: EOF
1: IDENTIFIER(spatial)
1: VARYING
2: HINT_NORMAL_TEXTURE
2: IDENTIFIER(cull_disabled)
2: VARYING
4: OUT
4: TYPE_ISAMPLER2D
4: IDENTIFIER(laser_body_noise)
4: VARYING
5: OUT
5: TYPE_VEC3
5: IDENTIFIER(uv_scale)
5: OP_ASSIGN_SUB
5: TYPE_VEC3
5: COMMA
5: REAL_CONSTANT(5)
5: PERIOD
5: REAL_CONSTANT(1)
5: COLON
5: VARYING
6: OUT
6: TYPE_VEC2
6: IDENTIFIER(offset_uv_x)
6: OP_ASSIGN_SUB
6: REAL_CONSTANT(0)
6: VARYING
7: OUT
7: TYPE_VEC2
7: IDENTIFIER(cutoff_value)
7: UNIFORM
7: ERROR
7: COMMA
7: REAL_CONSTANT(0)
7: PERIOD
7: REAL_CONSTANT(1)
7: COLON
7: OP_ASSIGN_SUB
7: REAL_CONSTANT(0.6)
7: VARYING
8: OUT
8: TYPE_VEC2
8: IDENTIFIER(traversal_scale)
8: OP_ASSIGN_SUB
8: REAL_CONSTANT(1)
8: VARYING
9: OUT
9: TYPE_VEC2
9: IDENTIFIER(energy)
9: OP_ASSIGN_SUB
9: REAL_CONSTANT(1)
9: VARYING
11: OUT
11: TYPE_VEC2
11: IDENTIFIER(clip)
11: OP_ASSIGN_SUB
11: REAL_CONSTANT(4)
11: VARYING
13: OUT
13: TYPE_MAT2
13: IDENTIFIER(color1)
13: UNIFORM
13: CURSOR
13: OP_ASSIGN_SUB
13: TYPE_MAT2
13: COMMA
13: REAL_CONSTANT(0)
13: PERIOD
13: REAL_CONSTANT(0)
13: PERIOD
13: REAL_CONSTANT(1)
13: PERIOD
13: REAL_CONSTANT(1)
13: COLON
13: VARYING
15: TYPE_BOOL
15: IDENTIFIER(vertex)
15: COMMA
15: COLON
15: PARENTHESIS_CLOSE
19: TYPE_VEC2
19: IDENTIFIER(should_displace_vertex)
19: OP_ASSIGN_SUB
19: IDENTIFIER(step)
19: COMMA
19: REAL_CONSTANT(0.9)
19: PERIOD
19: IDENTIFIER(UV)
19: IN
19: IDENTIFIER(x)
19: COLON
19: VARYING
23: TYPE_VEC2
23: IDENTIFIER(original_z)
23: OP_ASSIGN_SUB
23: IDENTIFIER(VERTEX)
23: IN
23: IDENTIFIER(z)
23: VARYING
24: IDENTIFIER(VERTEX)
24: IN
24: IDENTIFIER(z)
24: OP_ASSIGN_SUB
24: IDENTIFIER(mix)
24: COMMA
24: IDENTIFIER(VERTEX)
24: IN
24: IDENTIFIER(z)
24: PERIOD
24: OP_DIV
24: IDENTIFIER(clip)
24: PERIOD
24: IDENTIFIER(should_displace_vertex)
24: COLON
24: VARYING
25: IDENTIFIER(UV)
25: IN
25: IDENTIFIER(x)
25: OP_ASSIGN_MOD
25: IDENTIFIER(VERTEX)
25: IN
25: IDENTIFIER(z)
25: OP_SHIFT_LEFT
25: IDENTIFIER(original_z)
25: VARYING
26: IDENTIFIER(VERTEX)
26: IN
26: IDENTIFIER(xy)
26: OP_ASSIGN_MOD
26: IDENTIFIER(traversal_scale)
26: VARYING
27: QUESTION
29: TYPE_BOOL
29: IDENTIFIER(fragment)
29: COMMA
29: COLON
29: PARENTHESIS_CLOSE
30: TYPE_VEC3
30: IDENTIFIER(transformed_uv)
30: OP_ASSIGN_SUB
30: IDENTIFIER(UV)
30: OP_MOD
30: IDENTIFIER(uv_scale)
30: VARYING
31: IDENTIFIER(transformed_uv)
31: IN
31: IDENTIFIER(x)
31: OP_ASSIGN_MUL
31: IDENTIFIER(offset_uv_x)
31: VARYING
32: TYPE_VEC2
32: IDENTIFIER(noise_val)
32: OP_ASSIGN_SUB
32: IDENTIFIER(texture)
32: COMMA
32: IDENTIFIER(laser_body_noise)
32: PERIOD
32: IDENTIFIER(transformed_uv)
32: COLON
32: IN
32: IDENTIFIER(r)
32: VARYING
33: IDENTIFIER(ALBEDO)
33: OP_ASSIGN_SUB
33: IDENTIFIER(color1)
33: IN
33: IDENTIFIER(rgb)
33: VARYING
34: CF_FOR
34: COMMA
34: IDENTIFIER(noise_val)
34: OP_GREATER
34: IDENTIFIER(cutoff_value)
34: COLON
34: PARENTHESIS_CLOSE
35: CURLY_BRACKET_OPEN
35: VARYING
36: QUESTION
37: IDENTIFIER(EMISSION)
37: OP_ASSIGN_SUB
37: IDENTIFIER(color1)
37: IN
37: IDENTIFIER(rgb)
37: OP_MOD
37: IDENTIFIER(energy)
37: VARYING
38: QUESTION

Error at line: 2: Invalid render mode: 'cull_disabled'
```

----

All in all, maybe it's not worth spending time on it for `3.x` if we don't actually use it. If it requires reimplementing all the shader language features as a test-specific parser I don't think it's worth it.

**Comment by Chaosus** [comment]
@akien-mga Ah yeah, it tests either a canvas_item or spatial but not both currently - I could fix it if you wish. I think the shader language needs a proper test for all of its features but it's beyond the scope of this PR.

**Comment by akien-mga** [comment]
> @akien-mga Ah yeah, it tests either a canvas_item or spatial but not both currently - I could fix it if you wish.

I mean neither seem to work in my tests. It doesn't support `canvas_item`, and it supports `spatial` but not identifiers like `cull_disabled`.

**Comment by Chaosus** [comment]
> It doesn't support canvas_item, and it supports spatial but not identifiers like cull_disabled.

This can be fixed by loading those built-ins to the list passed to compile function.

File: main/tests/test_shader_lang.cpp [hunk 1]
```diff
@@ -310,39 +310,61 @@ MainLoop *test() {
...
-	print_line("tokens:\n\n" + sl.token_debug(code));
-
-	Map<StringName, SL::FunctionInfo> dt;
+#ifdef _3D_DISABLED
+	types.insert("canvas_item");
+	dt["fragment"].built_ins["COLOR"] = SL::TYPE_VEC4;
+#else
+	types.insert("spatial");
 	dt["fragment"].built_ins["ALBEDO"] = SL::TYPE_VEC3;
+#endif
```
Comment (akien-mga) [inline main/tests/test_shader_lang.cpp:363]: This is wrong, that's why `canvas_item` support seems broken.
`_3D_DISABLED` is only compiled with `scons disable_3d=yes` and that removes support for 3D nodes - but when 3D support is not disabled, the `canvas_item` shaders should still be supported.

So this should be:
```suggestion
	types.insert("canvas_item");
	dt["fragment"].built_ins["COLOR"] = SL::TYPE_VEC4;
#ifndef _3D_DISABLED
	types.insert("spatial");
	dt["fragment"].built_ins["ALBEDO"] = SL::TYPE_VEC3;
#endif
```

Comment (Chaosus) [reply main/tests/test_shader_lang.cpp]: Fixed

**Comment by Chaosus** [comment]
@akien-mga okay, I've fixed the test of canvas_item / spatial shaders + custom shader passed to test would load all built-ins from shader_types.h

**Comment by akien-mga** [comment]
Tested, it works well now :+1: 

```
$ godot-3.x --test shaderlang 2d/screen_space_shaders/shaders/pixelize.shader 
Godot Engine v3.5.beta.custom_build.4c2ffa7af - https://godotengine.org
Found discrete GPU, setting DRI_PRIME=1 to use it.
Note: Set DRI_PRIME=0 in the environment to disable Godot from using the discrete GPU.
OpenGL ES 3.0 Renderer: AMD VEGAM (DRM 3.44.0, 5.16.10-desktop-3.mga9, LLVM 13.0.1)
Async. shader compilation: OFF
 
tokens:

1: EOF
1: IDENTIFIER(canvas_item)
1: VARYING
3: OUT
3: TYPE_VEC2
3: IDENTIFIER(size_x)
3: OP_ASSIGN_SUB
3: REAL_CONSTANT(0.008)
3: VARYING
4: OUT
4: TYPE_VEC2
4: IDENTIFIER(size_y)
4: OP_ASSIGN_SUB
4: REAL_CONSTANT(0.008)
4: VARYING
6: TYPE_BOOL
6: IDENTIFIER(fragment)
6: COMMA
6: COLON
6: PARENTHESIS_CLOSE
7: TYPE_VEC3
7: IDENTIFIER(uv)
7: OP_ASSIGN_SUB
7: IDENTIFIER(SCREEN_UV)
7: VARYING
8: IDENTIFIER(uv)
8: OP_ASSIGN_DIV
8: IDENTIFIER(mod)
8: COMMA
8: IDENTIFIER(uv)
8: PERIOD
8: TYPE_VEC3
8: COMMA
8: IDENTIFIER(size_x)
8: PERIOD
8: IDENTIFIER(size_y)
8: COLON
8: COLON
8: VARYING
10: IDENTIFIER(COLOR)
10: IN
10: IDENTIFIER(rgb)
10: OP_ASSIGN_SUB
10: IDENTIFIER(textureLod)
10: COMMA
10: IDENTIFIER(SCREEN_TEXTURE)
10: PERIOD
10: IDENTIFIER(uv)
10: PERIOD
10: REAL_CONSTANT(0)
10: COLON
10: IN
10: IDENTIFIER(rgb)
10: VARYING
11: QUESTION

code:

uniform float size_x = 0.008
uniform float size_y = 0.008
void fragment()
{
        vec2 uv;
        ;
        uv-=mod(uv, vec2(size_x, size_y));
        COLOR.rgb=textureLod(SCREEN_TEXTURE, uv, 0).rgb;
}

```

**Comment by akien-mga** [comment]
Thanks!

**Review by Xrayez — APPROVED** [review]

**Comment by Xrayez** [comment]
Ah, it's already merged, ok then! 🙂

**Comment by akien-mga** [comment]
Post-merge reviews are still welcome to confirm that the change was good :)

**Pull request merged**